Variable sensitivity domain

martin-cs opened the pull request:
This PR merges in several years of work on a new domain for the abstract interpreter.  It should solve a number of the long-standing problems with soundness and precision.  The key feature of the domain is that the abstraction used for each type can be selected at run-time, allowing a wide range of scalability / precision trade-offs and easy specialisation without having to re-write all of the array, field, pointer sensitivity, etc.

1. This is the work of a lot of people, notably @thk123 @chrisr-diffblue @hannes-steffenhagen-diffblue @danpoe @NlightNFotis and others.  I have tried to preserve authorship where-ever possible.

2. I don't think it is reasonable to try to review this in the normal way, nor do I think it is viable to split it up further.  Whether it gets merged or not has to be a policy question really.  Over the years this has been going on I have asked @peterschrammel and @tautschnig about this as an approach and they seemed supportive because...

3. The PR should have very little impact on existing functionality, it is enabled with an option in goto-analyzer.  It has also been deployed in at least two industrial applications.

4. There is still work to be done in tidying up the code but I think it can be and would be best done as PRs to it after merging.  I will try to do some during the discussion / review.

5. This includes https://github.com/diffblue/cbmc/pull/5202 and https://github.com/diffblue/cbmc/pull/5204 which were previous attempts at splitting this out a bit.

6. I don't know who is currently looking after
https://github.com/diffblue/cbmc/tree/variable-sensitivity-domain
and
https://github.com/diffblue/cbmc/tree/variable-sensitivity-with-get-function-pointers
( @chrisr-diffblue ? @danpoe ? @hannes-steffenhagen-diffblue ? )
but this PR includes the majority of the content of these.  I have a branch of the remainder and am happy to work with people to rebase that on top of this PR.

Let me know what you think!

Commit:
Added implementation for transition into/out of function calls

The transform function now handles function calls (copy the values of
the parameters into appropriately named abstract objects) and function
return (erase the values of the parameters from the domain.

Commit:
Adding tests for function calls

Some of the code is commented out to disable unsupported features from the tests

Disabling parts of the test dependent on pointer arithmetic

Commit:
Added sound support for opaque functions

Currently just havocs the domain if an unknown function is called.

Commit:
Adding reasonable behaviour for opqaue function calls

When we have a function call without a body, we could set everything to
top (this would be more correct). To keep consistent, we default at the
moment to the more reasonable, less sound behaviour of only topping
pointers passed in and the return value.

Updated the test description to now be able to validate the simple
assert.

Commit:
Moved the function call handling into a separate method for readability

Commit:
Support for topping the global values

NlightNFotis committed:
Implemented the sensitivity domain for the union abstract object.

Adjusted to fit with the modifications on the variable sensitivity domain

Commit:
Fixed bug when doing a merge write to a pointer

When a pointer was written through (using a dereference) but it was a
merging write (e.g. when in an array), we were not assigning the new
value back into the domain and the value was lost. Added this case to
the array of pointers tests.

For the two value array we get the wrong result for the final two
assertions as we do not havoc the domain when writing to a pointer which
we don't know what it points to.

Commit:
Assertion for varags

When the parmeters and the arguments don't match, either the function
was called with too many args (in which case the function must be
variadic) or too few, in which case the code isn't valid. Added
appropriate assertions and a test for a variadic function.

Commit:
Reworked the double dispatch of the merge

Rather than using a double dispatch hidden inside a macro in
abstract_object, each object now handles the merge for the types that it
supports.

Made clone return a pointer to constant abstract object to ensure
correct usage.

Commit:
Removing default copy constructors

In all cases the copy constructor was doing what the default copy
constructor would do. Remove to avoid confusion

Commit:
Replacing map based switch statments with chained if/elses

To improve readability, the map of lambdas is replaced with a chain of
if else statements in both the eval and write methods.

Commit:
Explained why top must be used as the else clause in eval

I attempted to change this to generating a abstract object directly from
the simplified expression. This caused the tests to fail in what
appeared to be a "better" way - that is the constant_propagator tests
were reporting more assignment simplifications. Specifically, it was
simplifying the return of the function to be a NONDET(int) rather than
the return of the function.

I've convinced the change is wrong and works by chance, if we could some
how have exprt with no children but wasn't a symbol, but also wasn't
really a constant, then this would be wrong. The optimisation is also
probably wrong since if the return assignment happened earlier, the
return value should be the same, not a new non-det.

Commit:
Don't make copies unnecessarily when merging

When we merge, we don't want to create a clone of the object if nothing
changes as this reduces the ability of the sharing map to optimize.

This simplifies the code for detecting whether a merge has changed
anything, as we can compare the pointers, if they are pointing at the
same thing, nothing has changed.

It complicates the code for the merges as we need to be careful not to
leak any abstract objects. We can't use the shared pointer wrapper as we
want to be able to use the this pointer (unless we can use
std::enable_shared_from_this?)

Commit:
Using enable_shared_from_this to eliminate naked pointers

Modify the merge to return shared pointers by replacing returns of this
with returns of `shared_from_this`. This simplifies the code for dealing
with when need to modify the result in dervied classes as we are no
longer responsible for freeing the pointer returned by the super class.

Commit:
Adding unit tests for merging of abstract objects

Commit:
Fixed bugs revealed by unit tests

The base merge should always result in either top or bottom (it is up
for dervied classes to clear the top flag if they are not top).

Correctly use the base class merge when appropriate in merging constant
values.

Commit:
Adding unit tests for constant abstract object.

Commit:
Adding tests for constant array

Refactored the merge maps method to use std::intersection as there was a
bug with empty maps caught by the unit tests.

Commit:
Adding unit tests for the struct abstract object

Commit:
Corrected merge methods for pointer and struct

Commit:
Rewriting the constant_abstract_value::merge using behaviour driven design

Commit:
Replaced assert with __CPROVER_assert

Use the CPROVER_assert as behaviour is consistent across platforms.

Commit:
Use Behaviour Driven Design testing for abstract object merge

Commit:
Struct and array AO unit tests using behaviour driven design

Removed some unnecessary debug information from the tests

Commit:
Corrected guard on specialised merge

Commit:
Made pointer to array name consistent

Commit:
Constant array reading from top handled correctly

Commit:
Completed test suite for full struct and constant array AOs

Commit:
Correctly set up the enviroment before running tests

In release builds, the top/bottom status of the enviroment was
undefined. This ensures the enviroment is set to top (i.e. an entry
point).

Commit:
Extracted out logic for when to use a 2 value merge

Previously each abstract object that implemented merge had to correctly
redirect merging top with anything or merging with a bottom.

This is now handled directly in the root merge by checking the three
cases where the base merge must be used.

If this is bottom must still be handled by derived classes as we must
create an abstract object of the same type.

Commit:
Improved comment in the LHS simplify

Commit:
Made ai_simplify perform a partial simplifcation

If the ai_simplify is not able to resolve the expression to a constant,
it tries to simplify recursively any child operations. For example,
given a int i that is know to be 1, and an unknown (top) array a, it
would convert a[i] to a[1].

Commit:
Fixed ai_simplify in variable sensitivity domain to use new interface

At some point this commit needs backdating as the ai_domain interface
has changed under our feet.

Commit:
Corrected lhs simplification

True means no simplification has occured so this should be reflected in
the lhs simplification

Commit:
Adding tests for simplification of lhs

Commit:
Fixed constant propogation test

An additional assign is simplified as the LHS side simplification now
works.

Commit:
Corrected code in base class and removed duplicate code

Commit:
Implementing a write_stack to represent writing to a addressable location

Deals with storing a stack to write to things like:

&a[1]
&s.comp
&a[x] + 1

By building a stack as parsing the expression tree.

It does not support pointer arithmetic outside the scope of an array
(e.g. within a struct)

This will be used to store pointer values.

Disabled unit tests since don't have the prerequistes in this branch

Commit:
Use the write_stack in the constant_pointer_abstract_object

Commit:
Modified the write stack to not throw away certain pointers

Pointers into arrays was creating a top stack even though the result was
valid. Corrected so didn't throw them away (need to look at the base of
the stack, not the top). Also made the tests verify that they don't
create a top stack when they shouldn't. Also made it impossible to get
the expression of a top stack.

Commit:
Enable the write stack for pointers to structs

This works correctly so we don't need to cautiously throw away stacks of
pointers to structs. The original concern was pointers to structs being
used as pointers to the first component, but the goto program introduces
a cast which means we don't create the pointer.

Added a unit test to document this behavior.

Commit:
Enabling some of these assertions in tests related to arrays

We have to turn on pointer sensitivity or otherwise when evaluating
these expressions we won't get the correct write stack.

Commit:
Add initial output functions for all sub-classes except pointer.

Commit:
Add method to update last_written_locations and used it provisionally in ASSIGN.

Commit:
Experimental helper functions and an attempt to note destroy information for derefs, but do destroy (and correct!) information for other LHS assigns.

Commit:
Initial merge implementation, some new helper functions (to be moved to cpp and pruned), some notes and disabled removal of TOP elements.

Commit:
Add erase methods and pointer handling.

Commit:
Brief auxiliary function tidy-up

Commit:
Finished prior to clean.

Commit:
Implemented "last written location" tracking abstract objects

Each abstract_objectt now has an std::set<goto_programt::const_targett> ("last write locations") data structure attached to it.

For each assign instruction, whenever an object is assigned to a value, the last write location set for that object is cleared, and updated to contain the program location (specifically the const_targett.)

Furthermore, if the object is not a simple symbol (e.g. struct, array), then the high-level symbol is also updated to represent modification. (unless it's a pointer write, in which case only modified objects are updated.)

For writes where multiple objects are copied 'simultaneously' (within a single GOTO instruction, for example int a[2] = { 3, 4 } then all sub-elements of the written object are also updated.  So in this case, given current location is X, then last_written_locations will appoximately map a->X, a[0]->X, a[1]->X.  This applies recursively.

Merges perform an additional union between objects' last write location sets.

The last_written_locations information is automatically outputted if output() is called, it is represented by a @ [X] or a @[X, Y], etc.

Commit:
Correction of Thomas's comments.

Commit:
Extracted the output method for locationst (into environment class, discuss)

Commit:
To provide a fix for recursive assignments.

Recursion into the sub-element should not occur on symbol_exprt updates, this will 'clobber' all sub-elements and destroy all useful data.

Commit:
Initial set of tests with manual human inspection.   Moved static output method to abstract_object and using output in appropriate sub-classes

Commit:
Correction of formatting (function call over multiple lines).

NlightNFotis committed:
First version of the function that searches two environments for modified symbols.

NlightNFotis committed:
[WIP] Added more tests.

Commit:
Temporary commit for using the modified values

Commit:
Implementing a write_stack to represent writing to a addressable location

More random changes

Commit:
Another temp checkin of where we've got to

Commit:
Adding get_modified_symbols to other domainTs

Commit:
Restored the correct modified symbols

We can now see reliably which symbols have been modified at the end of a
function call.

NlightNFotis committed:
[WIP] Added the restore domain function.

Commit:
Use the restore_domain function to do context sensitive func return

NlightNFotis committed:
Added a test for the variable sensitivity assign aware merge, for the function pointer_function.

NlightNFotis committed:
Added more tests for variable sensitivity assign aware merge for function simple_test.

NlightNFotis committed:
Added more tests for variable sensitivity assign aware merge for function global_test.

NlightNFotis committed:
Added a basic file for array tests. Need thomas to config its validity before I author the test.desc

NlightNFotis committed:
Added a struct c test file, and made some changes to the array testing code.

NlightNFotis committed:
Added some changes that thomas made on my machine to test arrays modified by functions.

NlightNFotis committed:
Added the test.desc files for tests of arrays and structs.

Commit:
Fixing comment

Commit:
Correctly escaping braces

Commit:
Delete incorrect comment and fixing whitespace bug

Commit:
Update the last written location for structs

When updating a component in a struct container, we must update the last
written location for the struct container itself.

Commit:
Fixed typo introduced in test

Commit:
Add any elements that are only in the second map since this is modified

Commit:
Updated array test to test exactly what we want

NlightNFotis committed:
Fixed the struct test.desc

Commit:
Three way merge initial checkin

NlightNFotis committed:
Added tests for goto after function.

NlightNFotis committed:
Added tests for the non-terminating function case.

Commit:
Reworked three way merge that works when enabled and not

Commit:
Tidied up the changes introduced by context aware merging

chrisr-diffblue committed:
Correctly identify bitfield constants as CONSTANT values

Previously constant bitfield values were identified as TWO_VALUE by
default, meaning they became TOP immediately. This change makes them
be correctly identified as constants and hence have their values tracked
appropriately.

chrisr-diffblue committed:
Driveby cleanup to keep Lint happy

Commit:
Tweak to cope with missing utility function.

Commit:
Fix up the last written location sensitivity tests.

hannes-steffenhagen-diffblue committed:
Allows simplification of expressions containing non-const

Previously, abstract_objectt::expression_transform returned top
whenever one of the subexpressions of an operator was non-const,
which didn't allow simplification in situations like
int x = 0;
int y = nondet_int();
int z = x * y;

The old behavior would've replaced x * y with 0 * y,
with this change it will correctly simplify it to 0 instead.

chrisr-diffblue committed:
Factor out 'is_modified' logic into to the abstract_object implementations

Instead of having the abstract_environment have to know details of the
abstract_objects (e.g. last_written_location) in order to decide what
has been modified between two environments (e.g. at a 3-way merge), make
the individual implementation of abstract_object handle that logic.

src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp

chrisr-diffblue committed:
Add a common interface for read/write operations on abstract_objectt's

chrisr-diffblue committed:
Avoid direct manipulation of abstract_objectt::top

chrisr-diffblue committed:
Factor out last_written_location into it's own abstract_objectt

chrisr-diffblue committed:
Refactor and rename update_last_written_locations to be more generic

Two main changes:
  * update_last_written_locations is renamed to update_location_context
  * the update_sub_element functions/mechanism is replaced with a more
    generically useful visitor mechanism

chrisr-diffblue committed:
Fixup variable-sensitivity unit tests to build after the giant rebase.

chrisr-diffblue committed:
Use sharing_mapt instead of std::map in full_struct_abstract_object

Commit:
Convert assertions to invariants.

Commit:
Fix a crashing bug with type comparison.

Commit:
Crashing test cases due to pointer arithmetic issues.

Commit:
Test case for the evaluation of (false && x) in the variable sensitivity domain.

hannes-steffenhagen-diffblue committed:
Use sharing_mapt for abstract environment

Commit:
Use sharing_mapt in constant_array_abstract_objectt

chrisr-diffblue committed:
Fix a silly double call of visitor.visit() in full_struct_abstract_objectt

Commit:
Remove deprecated call to negate.

hannes-steffenhagen-diffblue committed:
Avoid unnecessarily breaking sharing

hannes-steffenhagen-diffblue committed:
Adjust float expression in variable sensitivity domain

chrisr-diffblue committed:
CLEANUP: Correct indentation

chrisr-diffblue committed:
Record struct initialisation when field sensitivity is turned on

Currently, full_struct_abstract_object does not properly track static
initialisation of a struct value. This commit ensures that when a struct
is initialised with a static value, the appropriate field values are also
tracked.

chrisr-diffblue committed:
Add data-dependency-context to variable sensitivity as commandline option

This commit adds an extra 'sensitivity' to the variable sensitivity domain
that enables it to track data-dependencies as an additional context on
'abstract_objectt's

chrisr-diffblue committed:
REFACTOR: Rename dependency_context_abstract_object to write_location_context

chrisr-diffblue committed:
CLEANUP: Remove unneeded include of full_struct_abstract_object.h

chrisr-diffblue committed:
CLEANUP: Correct authorship of dependency_context_abstract_object

chrisr-diffblue committed:
REFACTOR: Refactor variable sensitivity write_location_context

chrisr-diffblue committed:
CLEANUP: Make constant_array_abstract_object match full_struct_abstract_object

Both constant_array_abstract_object and full_struct_abstract_object have
similar logic to handle their additional sensitivity over their base classes
(array_abstract_object and struct_abstract_object) but their two implementation
code layouts have diverged quite a bit. This patch makes the two implementations
have a much closer layout to each other, which helps with comparing the
two implementations when debugging.

chrisr-diffblue committed:
Add variable sensitivity data dependency graph

chrisr-diffblue committed:
CLEANUP: Move dump_map utility functions to the correct .cpp file

hannes-steffenhagen-diffblue committed:
Try all rounding modes when rounding mode is unknown

For variable sensitivity domain. If we don't know the rounding mode
(i.e. it's not a constant known value), we can still often
evaluate expressions involving floating points if the results
don't change based on floating point (which is going to be the case
with floating point operations that have exact results or comparisons)

chrisr-diffblue committed:
[DO NOT MERGE] Track which expressions give rise to data dependencies

This is a customer special feature which tracks which expressions are the
source of any data dependencies, displaying those expressions as part of the
output. JSON output also includes the 'certainty' level of the dependency.
At the moment, this is a hard coded/enabled feature. It should be made
controlled by a command line option before it gets considered for merging, but
PR'ing anyway for review comments.

hannes-steffenhagen-diffblue committed:
Add floating point simplifcation test

hannes-steffenhagen-diffblue committed:
Fixups post-rebase

chrisr-diffblue committed:
Avoid breaking sharing in usage of sharing_mapt in full_struct_abstract_object

chrisr-diffblue committed:
Avoid breaking sharing in usage of sharing_mapt in constant_array_abstract_object

hannes-steffenhagen-diffblue committed:
Add regression tests for interval abstract values

danpoe committed:
Add data dependency context and variable sensitivity dependence graph to Makefile

danpoe committed:
Remove stray ; in context abstract object

hannes-steffenhagen-diffblue committed:
Make constant_array_abstract_object::eval_index virtual

Subclasses may need to override this behaviour

hannes-steffenhagen-diffblue committed:
Move interval value constructors to .cpp

hannes-steffenhagen-diffblue committed:
Move interval value constructors to .cpp

hannes-steffenhagen-diffblue committed:
Add interval arrays to variable sensitivity

hannes-steffenhagen-diffblue committed:
Clang format

NlightNFotis committed:
Fix interval abstract value constructor

Set interval abstract value to top or bottom if
the contained interval is top or bottom.

NlightNFotis committed:
Fix infinite loop in interval values

Make sure that the same value can't be merged more than
ten times. This is because in loops intervals kept changing
so we never reached a fixed point.

danpoe committed:
Moved rounding mode aware arithmetic to constant abstract value

danpoe committed:
Calling eval_expression on top abstract object

danpoe committed:
Mark failing dependence graph test as FUTURE

danpoe committed:
Escape curly braces in variable sensitivity float regression test

danpoe committed:
Add struct case to abstract environment eval

danpoe committed:
Basic interval arithmetic

danpoe committed:
Interval arithmetic regression tests

danpoe committed:
use eval to evaluate binary operators in interval abstract object

danpoe committed:
Typecast from ID_bool to integer types

hannes-steffenhagen-diffblue committed:
Return top in case of min/max indices in interval arrays

There's little else useful we can do in these cases.
Note: It might be worth to also return top if
the index is (or might be) out of bounds for the array type,
but this is inconsistent with how it currently works
for constant_array_abstract_objectt

danpoe committed:
Addressing review comments

hannes-steffenhagen-diffblue committed:
Exit loops in interval arrays early if result becomes top

hannes-steffenhagen-diffblue committed:
Use from_integer instead of constant_exprt constructor

danpoe committed:
Adapt dependence graph regression test to new output format

hannes-steffenhagen-diffblue committed:
Add function to get statistics about variable sensitivity domain

Currently, only number of interval objects and number of single-value intervals
are tracked, but this could easily be changed

chrisr-diffblue committed:
Fixups to enable building after rebase

hannes-steffenhagen-diffblue committed:
Add counts of various object types to environment statistics

hannes-steffenhagen-diffblue committed:
Add an estimation of memory usage to object statistics

This is an underestimation because it doesn't take into account
memory allocated for ireps and some data structures (like the contexts
for context_abstract_object)

chrisr-diffblue committed:
Remove INVARIANT in full_struct_abstract_object copy constructor

There are times during the process of writing to a struct that
the copy constructor gets called (via mutable_clone) while the
source object is in an 'in-flight' state which gets discarded by
the time the full write actions have been completed. Leaving the
INVARIANT check in-place causes some writes to fail because of the
transiently inconsistent state of the full_struct_abstract_object.

chrisr-diffblue committed:
Add support for bit field and enums intervals.cpp

chrisr-diffblue committed:
Make variable-sensitivity dependence graph produce control dependencies

Previously the variable sensitivity domain dependence graph only produced
data dependency information. This patch copies out the control dependency
code from the original dependency graph code and patches those into the
variable sensitivity dependence graph code. Note, this does *not* change
the algorithm used to calculate the control dependencies, it just means
we can calculate both control dependencies (using the original algorithm)
and the data dependencies using the new variable sensitivity based algorithm
in a single pass of the tools.

chrisr-diffblue committed:
Avoid creating interval_abstract_valuet's for float types

Currently interval_abstract_valuet only supports integral types,
but the object factory was attempting to create them for float types
anyway. This patch ensures that for float types we fall back to
creating constant_abstract_valuet instead.

chrisr-diffblue committed:
Add a basic implementation of to_constant for interval_abstract_valuet

chrisr-diffblue committed:
Crude ability to expression_transform mixed interval and non-intervals

chrisr-diffblue committed:
Added regression test for variable sensitivity intervals with mixed float/bool expressions

chrisr-diffblue committed:
Include function call arguments in variable sensitivity dependency graph

hannes-steffenhagen-diffblue committed:
Disable currently broken dependence graph tests

hannes-steffenhagen-diffblue committed:
Disable interval float test

hannes-steffenhagen-diffblue committed:
Fix variable sensitivity unit test

danpoe committed:
Add control dependencies from function call to function entry point for new dependency graph

chrisr-diffblue committed:
Fix write_stack to correctly complete types for index_exprt expressions

chrisr-diffblue committed:
Avoid breaking fixpoint when merging data_dependency_context

When merging two data_dependency_context abstract objects, ensure a new
abstract object is returned iff the dependencies have actually been updated.
Previously a new result object was getting unconditionally returned, even if
its data did not change. This was causing fixpoint to fail to terminate.

danpoe committed:
Add specialised ai domain methods for variable sensitivity dependence graph

danpoe committed:
Control dependencies of function body instructions to function calls for variable sensitivity dependence graph

chrisr-diffblue committed:
Include GOTO guard expressions in variable sensitivity data dependencies

danpoe committed:
Fix variable sensitivity dependence graph regression test

chrisr-diffblue committed:
Ensure constant_array_abstract_object is in a valid state after calling make_top

hannes-steffenhagen-diffblue committed:
Initialise is_bottom in abstract_environmentt

chrisr-diffblue committed:
Fix variable sensitvity to build with XCode 9

chrisr-diffblue committed:
Track only immediate dependencies in variable sensitivity dependence graph.

Previously the variable sensitivity dependence graph would track both immediate
and transitive dependencies. Transitive dependencies are not needed and can be
calculated by other means. This change brings variable sensitivity dependence
graph into line with the behaviour of the old dependence graph code.

chrisr-diffblue committed:
Unconditional control dependency fix

chrisr-diffblue committed:
Fix dependence graph precondition failure in VSB dependence graph

chrisr-diffblue committed:
Fixup incorrect itterator comparison in data_dependency_context

Because the entries in dependency sets are locationt's (and therefore
itterators form potentially different collections) it's not safe to directly
compare dependency sets.

chrisr-diffblue committed:
Add a module_dependencies.txt to variable sensitivity domain

chrisr-diffblue committed:
Record data dependencies for functions with missing bodies

chrisr-diffblue committed:
Avoid setting domain to TOP when calling functions with const pointer arguments

Commit:
Merge when reading from array with unknown index.

Commit:
Tools to build intersection of intervals satisfying bool expressions

Commit:
Added regression test for PR #295

petr-bauch committed:
Fix interval construction

when the expression to be represented is neither constant interval nor a
relation. We simply return TOP interval. Includes a test that was failing.

chrisr-diffblue committed:
Support interval expressions containing ?: operations

Commit:
Interval values

Commit:
Various code style fixes in intervals code

Commit:
Handle typecasts in constant_interval_exprt

Commit:
Extend interval multiplication for extremes

Commit:
Interval operations

Commit:
Move ai configuration to separate config object

This lets us reuse the logic of getting the ait
configuration from command line options.

Also, this includes a small simplification
of the printing logic

Commit:
Add option to print statistics about the VS domain

Commit:
Apply bandaid fix to variable sensitivity

The C++ standard leaves comparing iterators from different containers against
each other as undefined behaviour. We've been doing this with goto locations,
which let us run into standard library assertion failures during testing in
debug builds.

This "fixes" this by avoiding the usage of the standard comparison operators for
iterators in places that led to regression test failures.

Going forward, we should probably consider representing goto locations in a less
brittle manner, or at least provide a wrapper with sane comparison semantics for
the list iterators we're currently using.

owen-mc-diffblue committed:
Implement value sets

as a new abstract object.

owen-mc-diffblue committed:
Extend the interface with hashing and equality  for unordered-sets.

Commit:
If there is no context to wrap - don't wrap the objects

There is no value in wrapping an object with no additional information,
and there is currently some problems with wrapping

Commit:
Only enable location context if value set is turned off

There is currently some inconsistency between how the value set and the
interval domains handle wrapping. Specifically, the value set returns an
unwrapped object when eval'd, whereas the interval domain returns a
wrapped object. Since the rest of the code is compatible with the
interval suggestion, we temporarily disable wrapping if using value set

Commit:
Produce an error if value set is used with a context tracking argument

Commit:
Remove unusued boolean flag

The variable sensitivity domain is always on, if we're using the
variable sensitivity domain object factory

Commit:
Remove the way to set configuration options using only optionst

Commit:
Introduce some sensible config defaults

Commit:
Use the config to simplify the code

Commit:
Add tests for value-set domain

owen-mc-diffblue committed:
Make value set entries human-readable

Use from_expr instead of constant_expr::get_value() so we don't print hex values

owen-mc-diffblue committed:
Various miscellaneous fixes

Most of these are needed to make clang-format happy, or
the linter, or to make it compile on Windows

owen-mc-diffblue committed:
Delete write_stack unit test

It was written in 2017 but was #if 0-ed out, so it needs to be updated to compile now.
ADA-496 has been filed to fix this test up and reintroduce it.

NlightNFotis committed:
Fix test and add reference to github issue.

NlightNFotis committed:
Make variable_sensitivity_domaint::eval public from protected

hannes-steffenhagen-diffblue committed:
Add stub implementations for value set classes

hannes-steffenhagen-diffblue committed:
Use brace initialisation instead of parentheses

hannes-steffenhagen-diffblue committed:
Implement merge for value_set_abstract_valuet

hannes-steffenhagen-diffblue committed:
Implement output for value-set abstract values

hannes-steffenhagen-diffblue committed:
Implement value_set_abstract_valuet::to_constant

Commit:
clang-format everything

Commit:
Update test description for the move to __CPROVER_assert

Somehow these changes had got lost during development.

Commit:
Rather than change existing tests, create new test descriptions

A number of the tests for the old constants domain had been
converted to test the variable sensitivity domain.  This would
reduce the test coverage of the --constants domain.  So instead
new .desc files have been added so both can be tested on the
same input programs, allowing side-by-side comparison.

Commit:
Three way merge is not being added to ai_domaint so no override

Commit:
Update the use of exprts so that it is more type safe

Commit:
Disable stats collection until the necessary changes are merged to ait

Commit:
Replace call to a utility function that is not in the patch set

Commit:
Remove an unreachable fall back case

ai_domain_baset::merge_three_way_function_return has been removed
in favour of a simpler interface, so the call is removed.  At the
moment it is unreachable anyway so it is kind of a mute point.

Commit:
Another update for the change in how three_way_merge is integrated

Commit:
Port over some of the changes to dependence_graph to the VS version

Commit:
fixup the unreachable thing

Commit:
Tweak the update to VSDG

Commit:
Refactor the VS dependency graph analysis to use the domain factory

This is a direct port of 07f8696e1c8168b3e61b024841720473504a573b
which use(d|s) the factory interface to build domains with a
non-trivial constructor, in this case passing the location.  This
avoids creating invalid objects and doing the "lazy construction"
in get_state.

Commit:
Fix all of the linting errors

Commit:
Fix a very subtle bug in the initialisation of VSD domain objects

The default state of a domain on construction should be bottom.
This is documented in the ai_domaint interface.
As the variable_sensitivity_domaint delegates the handling of bottom
to the abstract_environment, it must be initialised to bottom in the
default case.

Commit:
Integrate VSD so that it can be called from goto-analyzer

The command-line flags are backwards compatable with older versions
of this code where possible.

Commit:
Update the regression tests for the new arguments

Commit:
Re-enable dependency graph tests that were disabled at some point

Commit:
Tidy up, fix and update the constant propagation tests (VSD and not)

Commit:
Make tests independent of the exact line numbers for assertions

Commit:
Create a new abstract interpreter that does a three-way merge

Three-way merging is a VSD specific feature.  On function call return,
rather than just merging in the final state of the function to the
return context, it computes the difference between the first and last
states and then merges that into the calling state.  The key difference
is that if there are multiple calling contexts, their state will be
merged and normally picked up.  For example;

global = 1;                  global = 2;
f00();                       f00();

The start of f00 must have global as the merge of the two (i.e. top if
constant) and this will be merged into the return state -- even if
f00 obviously does not touch global.  Three-way merge only modifies
the variables that change during f00 so in this case, global will be
unmodified in both contexts.

All of this only makes sense if there are multiple calling locations that
merge into the same history.  So if a fully context sensitive history
is used then there is no point using three-way-merge.

This commit also updates the tests that actually need this functionality.
They demonstrate some of the other advantages of this technique.

Commit:
Enable three-way merge for the tests that need it

Commit:
Update line numbers for domains that track changes

Commit:
Use three-way merge in the VSD dependency graph

This requires a few legacy contortions similar to the ones
done in dependency graph.

Commit:
Update the VSD dependency graph tests for line number changes

These are due to new global variables being added to
__CPROVER_initialise

martin-cs commented:
The unit tests currently don't compile (oops!).  I note that "AWS CodeBuild us-east-1 (cbmc-windows-cmake)" passes; is it supposed to skip / ignore the unit tests?

peterschrammel commented:
Happy to review if you make the monster pass CI first.

peterschrammel commented:
> is it supposed to skip / ignore the unit tests?

It probably isn't. @hannes-steffenhagen-diffblue ?

martin-cs commented:
@peterschrammel Thank you, I think it is just the unit tests which are outstanding.  All the regression tests and linting, formatting, etc. all pass.  I will update the usage of `exprt` in the unit tests and ping you.

Commit:
Update unit tests for the improved, type safe(r) exprts

hannes-steffenhagen-diffblue commented:
@martin-cs @peterschrammel The cmake windows build is currently not running any tests actually. This is a long outstanding bug which is one of the things we're hoping to "fix" with our "move CI to github actions" work.

Commit:
Tweak the VSDG to fix a clang specific warning about get_state

martin-cs commented:
@hannes-steffenhagen-diffblue cool, as long as this isn't news to you!

martin-cs commented:
@hannes-steffenhagen-diffblue you probably already know this but it looks like "AWS CodeBuild us-east-1 (cbmc-clang-3-8)" doesn't run any tests either.

Commit:
One of the new unit tests was omitted from the source list

The N_CATCH_TESTS check in the Makefile found this!

Commit:
Update unit test for the bug fix to abstract_environmentt

In a previous commit I fixed a bug where the default state
of a constructed abstract_environment was top rather than
bottom.  Correcting this meant this unit test fails as the
environment will only return objects which are bottom.

Also improve the tagging of the test.

Commit:
Correct a typo in the creation of an interval

Thankfully the names of the variables and the description
of the tests makes it clear that this was a typo.
Documentation FTW!

martin-cs commented:
@peterschrammel All of CI now passes, any kind of review would be most appreciated.

peterschrammel commented on regression/goto-analyzer-simplify/simplify-lhs-pointer-array-dereference/main.c:
```diff
@@ -0,0 +1,21 @@
+
+  // TODO: Currently writing to an offset pointer sets the domain to top
+  // so recreate the variables to reign the domain back in
+  // This will be addressed by diffblue/cbmc-toyota#118
```
:pick: I don't think this reference is publically accessible.

martin-cs replied:
Removed.

peterschrammel commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,808 @@
+
+ Purpose: Evaluate the value of an expression relative to the current domain
+
+\*******************************************************************/
```
❗ Outdated comment headers. This needs replacing - if not in this PR than immediately as a follow-up.

martin-cs replied:
I have reformatted all of them.

peterschrammel commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,808 @@
+ what kind of array abstraction it is).  So, as we find the variable
+ ('a' in this case) we build a stack of which part of it is accessed.
+
+ As abstractions may split the assignment into multiple write (for
```
:pick:
```suggestion
 As abstractions may split the assignment into multiple writes (for
```

martin-cs replied:
Fixed.  Thanks.

peterschrammel commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,808 @@
+    // to be none of that.
+    if(s.id() != ID_symbol)
+    {
+      throw "invalid l-value";
```
❓ What is an `i-value`?

martin-cs replied:
? Is this a font issue? The source says `l-value` as in https://en.wikipedia.org/wiki/Value_(computer_science)#lrvalue

peterschrammel commented on src/analyses/variable-sensitivity/abstract_enviroment.cpp:
```diff
@@ -0,0 +1,808 @@
+
+  /* TODO : full implementation here
+   * Note that this is *very* syntax dependent so some normalisation would help
+   * 1. split up conjucts, handle each part separately
```
```suggestion
   * 1. split up conjuncts, handle each part separately
```

martin-cs replied:
Fixed.

peterschrammel commented on src/analyses/variable-sensitivity/abstract_enviroment.h:
```diff
@@ -0,0 +1,101 @@
+    const namespacet &ns);
+  virtual bool assume(const exprt &expr, const namespacet &ns);
+
+  virtual abstract_object_pointert write(
```
When converting the docs, please move interface docs into the header files. This would make it much easy to understand what a class is doing.

martin-cs replied:
Done.

peterschrammel commented on src/analyses/variable-sensitivity/abstract_object.h:
```diff
@@ -0,0 +1,404 @@
+/*******************************************************************\
+
+ Module: analyses variable-sensitivity
```
Please convert to doxygen.

martin-cs replied:
Done.

peterschrammel commented on src/analyses/variable-sensitivity/abstract_object.h:
```diff
@@ -0,0 +1,404 @@
+ abstract_objectt is the top of the inheritance heirarchy of objects
+ used to represent individual variables in the general non-relational
+ domain.  It is a two element abstraction (i.e. it is either top or
+ bottom).  Within the hierachy of objects under it, child classes are
```
```suggestion
 bottom).  Within the hierarchy of objects under it, child classes are
```

martin-cs replied:
Fixed.

peterschrammel commented on src/analyses/variable-sensitivity/abstract_object.h:
```diff
@@ -0,0 +1,404 @@
+   * compare against
+   * \return true if 'this' is considered to have been modified in comparison
+   * to 'before', false otherwise.
+   */
```
Use `///` stile of doxygen comments.

martin-cs replied:
I thought this was an acceptable style?  I thought it was used elsewhere in the code?

peterschrammel commented on src/analyses/variable-sensitivity/array_abstract_object.cpp:
```diff
@@ -0,0 +1,211 @@
+   merging_write - ?
+
+ Outputs: The struct_abstract_objectt representing the result of writing
+          to a specific component. In this case this will always be top
```
❓ The value doesn't seem to be `top` when you pass in `bottom`?

martin-cs replied:
The code is wrong but the way it is used will (I think) mean this case should never come up.  There is work to do on write stacks and writing to an index so I will add this to the TODO for that if it's OK with you?

peterschrammel commented on src/analyses/variable-sensitivity/constant_array_abstract_object.cpp:
```diff
@@ -0,0 +1,580 @@
+    }
+
+    // try to write to all
+    // TODO(tkiley): Merge with each entry
```
❓ Not sure about this comment. Is there a test that exposes a related problem? Or it should at least be explained why the current code is wrong.

martin-cs replied:
It's not that the current code is wrong, more that it could be improved.  If your index is not resolvable to a constant then there are a number of possible approaches.  Again, it is on my "to improve" list.

peterschrammel commented on src/analyses/variable-sensitivity/constant_array_abstract_object.cpp:
```diff
@@ -0,0 +1,580 @@
+  array_abstract_objectt::get_statistics(statistics, visited, env, ns);
+  shared_array_mapt::viewt view;
+  map.get_view(view);
+  for(auto const &object : view)
```
⛏️ Preferred for consistency with the rest of the codebase:
```suggestion
  for(const auto &object : view)
```

martin-cs replied:
Fixed.

peterschrammel commented on src/analyses/variable-sensitivity/abstract_object.h:
```diff
@@ -0,0 +1,404 @@
+class abstract_environmentt;
+class namespacet;
+
+#define CLONE                                                                  \
```
That's quite a generic macro name for something relatively specific.

martin-cs replied:
It is following the https://en.wikipedia.org/wiki/Prototype_pattern .  There is another good question over whether it is really needed.  Once the factory is refactored it should be clearer if and how it is used.

peterschrammel commented on src/analyses/variable-sensitivity/data_dependency_context.cpp:
```diff
@@ -0,0 +1,366 @@
+
+/**
+ * Helper function for abstract_objectt::abstract_object_merge to perform any
+ * additional actions after the base abstract_object_merge has completed it's
```
```suggestion
 * additional actions after the base abstract_object_merge has completed its
```

martin-cs replied:
Fixed.

peterschrammel commented on src/analyses/variable-sensitivity/full_struct_abstract_object.cpp:
```diff
@@ -0,0 +1,497 @@
+      struct_type_it->get_name(),
+      environment.abstract_object_factory(param_it->type(), *param_it, ns));
+    did_initialize_values = true;
+    ++struct_type_it;
```
:pick: put increment into for?

martin-cs replied:
Done.

peterschrammel commented on src/analyses/variable-sensitivity/interval_abstract_value.cpp:
```diff
@@ -0,0 +1,558 @@
+#endif
+}
+
+abstract_object_pointert interval_abstract_valuet::expression_transform(
```
💡  Consider breaking down this giant function.

martin-cs replied:
Fair point.  We do need to change how things are evaluated and the dispatch for that.  I hope that might simplify things a bit.

peterschrammel commented on src/analyses/variable-sensitivity/interval_array_abstract_object.h:
```diff
@@ -0,0 +1,55 @@
+    const namespacet &ns);
+
+protected:
+  CLONE
```
This macro looks really scary ;)

martin-cs replied:
Yeah, as above, I don't know if we can just remove it.

peterschrammel commented on src/analyses/variable-sensitivity/struct_abstract_object.h:
```diff
@@ -0,0 +1,65 @@
+    const namespacet &ns,
+    const std::stack<exprt> stack,
+    const exprt &specifier,
+    const abstract_object_pointert value,
```
❓ 
```suggestion
    const abstract_object_pointert &value,
```

martin-cs replied:
Fair point but it is a pointer so it shouldn't cost much.

peterschrammel commented on src/analyses/variable-sensitivity/struct_abstract_object.h:
```diff
@@ -0,0 +1,65 @@
+    const namespacet &ns,
+    const std::stack<exprt> &stack,
+    const member_exprt &member_expr,
+    const abstract_object_pointert value,
```
❓

martin-cs replied:
As above.

peterschrammel commented on src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp:
```diff
@@ -0,0 +1,255 @@
+\*******************************************************************/
+
+/// \file
+/// An abstract interpreter, based on the default recursive-interprocedural
```
No need to repeat this here.

martin-cs replied:
I am not sure I follow.  This class is one of the few that actually is an abstract interpreter.  Almost all of the functionality is inherited from other classes but it really is different from the rest of the code in this directory as it isn't really part of VSD.

peterschrammel commented on src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp:
```diff
@@ -0,0 +1,255 @@
+      ns);
+
+    // TODO : this is probably needed to avoid three_way_merge modifying one of
+    // it's arguments as it goes.  A better solution would be to refactor
```
```suggestion
    // its arguments as it goes.  A better solution would be to refactor
```

martin-cs replied:
Fixed.

peterschrammel commented on src/analyses/variable-sensitivity/three_way_merge_abstract_interpreter.cpp:
```diff
@@ -0,0 +1,255 @@
+  return new_data;
+}
+
+#if 0
```
Please document why this is commented out, or remove.

martin-cs replied:
Ooops :-\  Removed!

peterschrammel commented on src/analyses/variable-sensitivity/value_set_abstract_object.h:
```diff
@@ -0,0 +1,228 @@
+  ///   object of the right type if necessary.
+  /// \param new_values: potentially new set of values
+  /// \return the abstract object representing \p new_values (either 'this' or
+  ///   something new)
```
👍  Here the format is right.

peterschrammel commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.h:
```diff
@@ -0,0 +1,184 @@
+ Author: Thomas Kiley, thomas.kiley@diffblue.com
+
+There are different ways of handling arrays, structures, unions and
+pointers.  Our existing solution basically ignores them which is
```
Concretely refer to the module/class of the existing solution.

martin-cs replied:
Done.

peterschrammel commented on src/analyses/variable-sensitivity/variable_sensitivity_object_factory.h:
```diff
@@ -0,0 +1,254 @@
+      options.get_bool_option("pointers");
+
+    // This should always be on (for efficeny with 3-way merge)
+    // Does not work with value set
```
❓ Do we know why it doesn't work?

martin-cs replied:
The comment is by @thk123 so I fear the answer to that may have left with him.  I don't know and regard it as a "to fix".

peterschrammel commented on src/analyses/variable-sensitivity/write_location_context.cpp:
```diff
@@ -0,0 +1,351 @@
+
+/**
+ * Update the location context for an abstract object, potentially
+ * propogating the update to any children of this abstract object.
```
```suggestion
 * propagating the update to any children of this abstract object.
```

martin-cs replied:
Fixed.

peterschrammel commented on src/analyses/variable-sensitivity/write_location_context.cpp:
```diff
@@ -0,0 +1,351 @@
+ * \param update_sub_elements if true, propogate the update operation to any
+ * children of this abstract object
+ *
+ * \return a clone of this abstract object with it's location context
```
```suggestion
 * \return a clone of this abstract object with its location context
```

martin-cs replied:
Fixed.

peterschrammel commented on src/goto-analyzer/goto_analyzer_parse_options.h:
```diff
@@ -129,7 +130,17 @@ class optionst;
   "(non-null)" \
   "(constants)" \
-  "(dependence-graph)"
+  "(dependence-graph)" \
```
💡 `dependency-graph` seems to be more common.

martin-cs replied:
Yes but this is what the class in question is called and it has been a command-line option for some time.  I'm happy to change it but I don't think it should be part of this PR.

peterschrammel commented on unit/analyses/variable-sensitivity/abstract_object/merge.cpp:
```diff
@@ -0,0 +1,111 @@
+
+ Module: Unit tests for variable/sensitivity/abstract_object::merge
+
+ Author: DiffBlue Limited. All rights reserved.
```
❗ 
```suggestion
 Author: Diffblue Limited. 
```

martin-cs replied:
Fixed.

peterschrammel commented on unit/analyses/variable-sensitivity/constant_abstract_value/merge.cpp:
```diff
@@ -0,0 +1,580 @@
+
+ Module: Unit tests for constant_abstract_valuet::merge
+
+ Author: DiffBlue Limited. All rights reserved.
```
❗

martin-cs replied:
Fixed.

peterschrammel commented on unit/analyses/variable-sensitivity/constant_array_abstract_object/merge.cpp:
```diff
@@ -0,0 +1,702 @@
+
+ Module: Unit tests for constant_array_abstract_object::merge
+
+ Author: DiffBlue Limited. All rights reserved.
```
❗ (and all other files)

martin-cs replied:
Fixed.

peterschrammel commented on unit/analyses/variable-sensitivity/eval.cpp:
```diff
@@ -0,0 +1,100 @@
+// Copyright 2016-2020 Diffblue Limited. All Rights Reserved.
```
❗

martin-cs replied:
Fixed.

peterschrammel reviewed (commented):
A shallow review of 20KLOC to scan for red flags. The PR is in pretty good state code quality-wise - ie acceptable to be merged. The main concern are outdated documentation formats that need to be fixed so that the doxygen documentation is produced correctly. Also, some copyright headers don't follow the guidelines and must be fixed before merge.

Commit:
Remove reference to non-public issue tracker

Commit:
Convert comments to the new(er) doxygen format

Commit:
Reformat so that comments pass linting

Commit:
Apply Peter Schrammel's review comments

martin-cs commented:
@peterschrammel I need to move various doxygen comments into the headers but otherwise I should be done with the requested changes.  Let me know if there are more.

@chrisr-diffblue, @danpoe, @hannes-steffenhagen-diffblue : you know this code, any chance of a review?

Commit:
Move doxygen comments from .cpp to .h files

martin-cs commented:
@peterschrammel I have moved the doxygen comments to the headers.  Modulo fixing any issues doxygen has, I think I'm done.

martin-cs commented:
@peterschrammel Are the changes OK?

 @danpoe @hannes-steffenhagen-diffblue @smowton you are owners for this area, please can you say something about this.  I don't think an in-depth review is appropriate but some kind of comment would help.

@owen-mc-diffblue If I understand correctly, you were an owner under a different name that hasn't been updated in CODEOWNERS.  Please can you comment?

peterschrammel commented:
@martin-cs, lgtm modulo the doxygen issues: https://travis-ci.org/github/diffblue/cbmc/jobs/725557168

martin-cs commented:
Thanks @peterschrammel I will work through the doxygen issues once I have it set up on my machine.

Commit:
Fix issues and warnings raised by doxygen

martin-cs commented:
The doxygen issues should be finished.  I think it's just waiting on a stalled CI job.

Commit:
Update regression tests to cope with changes to initialisation

A recent PR has removed a couple of the in-built global
variables that are used for modelling the C standard library.
This has a knock on effect for the line numbers in the tests
of dependency graphs.

martin-cs commented:
@NlightNFotis I'm seeing build failures on MacOS.  But when I try to get the info about them I get:

"Error: We are currently unable to download the log. Please try again later. "

Any thoughts?  Did your PR fix everything?

martin-cs commented:
@peterschrammel @tautschnig @chrisr-diffblue @smowton I can't seem to access the build logs for a failing github action

`pull-request-checks / check-macos-10_15-cmake-clang (pull_request)`

any chance of some help?

peterschrammel commented:
@martin-cs, you should be able to see the log when you click on "Details" next to the failed job (provided that you are logged into github).

martin-cs commented:
More specifically,

https://github.com/diffblue/cbmc/pull/5472/checks?check_run_id=1379416806

says

```
Error: We are currently unable to download the log. Please try again later.
```

I've tried the procedure given in https://github.com/diffblue/cbmc/blob/develop/AccessingCodebuildLogs.md but it doesn't seem to work.

martin-cs commented:
Thanks @peterschrammel I am logged in and have tried that.  Any more suggestions?  Can you see it?

peterschrammel commented:
I can see it. You have all permissions to access it. So, either a temporary Github issue or a problem with your browser.

martin-cs commented:
Thanks @peterschrammel that's weird...  I will investigate.  What permissions do people need to see these?

peterschrammel commented:
It seems for public repos the only requirement is that you are logged in. No org membership or repo permissions required.

martin-cs commented:
Cool; have got it working (browser issue) thanks for the prompt support.

Commit:
s/const auto/const auto &/ in loops

This was inefficient but now causes build failures on OS X
as it uses -Werror -Wrange-loop-analysis

NlightNFotis commented:
This PR looks to be in a good state, and it's mergeable. Is there anything outstanding on it? If not, I can merge it now.

martin-cs commented:
@NlightNFotis  Thanks.  There are still quite a few refactorings, fixes, improvements, etc. that we will be making but TBH I would rather merge this and have these as separate PRs so they are easier to review.  The code in this PR works, has been deployed, has a higher test coverage than the rest of the project -- let's get it merged.  Plus it would be good to acknowledge everyone's hard work on this.

While we are on the subject, do you / @hannes-steffenhagen-diffblue / @chrisr-diffblue / @peterschrammel want / need any of the rest of the code from this branch?  I have a branch with it in, partially ready to be converted into PRs and I'm happy to work with people to do that but it is not a high priority if it is just me.

peterschrammel commented:
@martin-cs, it's favourable to get any (valuable) code that is lying around in branches in various forks into the shape of PRs eventually. It's shame when a lot of work has been done and then it rots and dies... and someone else starts reimplementing something similar...

martin-cs commented:
@peterschrammel agreed it is a question of priorities and time.  I could make them into draft PRs but they are unlikely to make mergable without someone else getting involved.

NlightNFotis commented:
@martin-cs I think there is value for the draft PRs to be made - if we can find some time we can shape them up to be more mergeable.

Obviously, I cannot give guarantees for when that is, as it depends on a lot of things, but if the PRs are there, they are much more likely to be picked up by someone.

thk123 commented:
😮  Never thought I'd see the day - can't let it go without commenting! 🎉 🎉 🎉

Hope you're all well

smowton commented:
Signs of the end times:
[x] POTUS refusing to leave the white house
[x] The variable-sensitivity domain has been merged
[ ] Rains of branston pickle

chrisr-diffblue commented:
This really should have been merged on 1st April....

hannes-steffenhagen-diffblue commented:
Hello @thk123 and @smowton, nice to hear from you! I hope you're all well too!

martin-cs commented:
@chrisr-diffblue ... I should have thought of that ... *sigh*

peterschrammel commented:
11/11 11:11 is the beginning of carnival...
